### DISCUSSION TITLE
Mapping between public keys and PeerIds

### DISCUSSION THREAD

**Discussion by laptou** [libp2p/rust-libp2p#2419] [Q&A] [answered]
Since `PeerId`s seem to be derived from a user's public key in the recommended usage, what happens when the same user wants to be active across multiple devices? I figured that I would just find a way to synchronize the user's key pair between devices, but I am not sure that having two nodes on the same network with the same `PeerId` is allowed. What should I do?

The protocols I am using specifically at the moment are mDNS, Kademlia, and Gossipsub.

**Comment by dvc94ch** [comment]
You can probably use identify to pass an additional user public key in one of the fields. You'll have to do the rest yourself

**Comment by rkuhn** [comment, marked as answer]
My suggestion is to not couple PeerId to users at all — a peer is a p2p network node, and as you say a single user may well use multiple nodes. These nodes need one unique identity each, even though they serve the same human. The associated PublicKey is an implementation detail of libp2p, I wouldn’t use it to represent the user either. A user should have their own SecretKey etc. which is completely separate from libp2p keys.

So you’ll need to implement a way on your application level for transferring a SecretKey between nodes of the same end user and then use that for encryption and signatures on the application data — in addition to all the transport layer encryption employed by libp2p.

**Reply by mxinden** [reply]
>  but I am not sure that having two nodes on the same network with the same PeerId is allowed.

Two online nodes with the same `PeerId` in the same network is not allowed.

> My suggestion is to not couple PeerId to users at all — a peer is a p2p network node, and as you say a single user may well use multiple nodes. 

Agreed with @rkuhn. Similar strategy is e.g. followed by https://github.com/paritytech/substrate.
